Remove `@inbounds` in tuple iteration

Keno opened the pull request:
LLVM can prove this inbounds and the annotation weakens the inferable effects for tuple iteration, which has a surprisingly large inference performance and precision impact. Add a test for both effects and codegen to make sure this doens't regress.

Keno commented:
Sigh, guess the final version of my PR wasn't actually enough to retain consitency here. I'll push an update.

aviatesk commented:
Guess the `:consistent`-cy is tainted here https://github.com/JuliaLang/julia/blob/2ca0981d29bfe6ebbe0497d456fea89410e5a539/base/compiler/tfuncs.jl#L2198-L2205

Keno commented:
Yes. I have most of a fix, but something is breaking, so I'll have to resume in the morning.

Keno committed:
Remove @inbounds in tuple iteration

LLVM can prove this inbounds and the annotation weakens the
inferable effects for tuple iteration, which has a surprisingly
large inference performance and precision impact.

Unfortunately, my previous changes to :inbounds tainting weren't
quite strong enough yet, because `getfield` was still tainting
consistency on unknown boundscheck arguments. To fix that, we
pass through the fargs into the fetfield effects to check if
we're getting a literal `:boundscheck`, in which case the
`:noinbounds` consistency-tainting logic I added in #48246
is sufficient to not require additional consistency tainting.

Also add a test for both effects and codegen to make sure this doens't regress.

Keno commented:
CI isn't running on this PR - not sure why. Let me just recreate it.
